src/handler/register_wallet.c

```diff
@@ -123,44 +130,39 @@ void handler_register_wallet(dispatcher_context_t *dc, uint8_t protocol_version)
         return;
     }
 
-    if (!ui_display_register_wallet(dc, &wallet_header, (char *) policy_map_descriptor)) {
-        SEND_SW(dc, SW_DENY);
-        ui_post_processing_confirm_wallet_registration(dc, false);
-        return;
-    }
-
     uint32_t master_key_fingerprint = crypto_get_master_key_fingerprint();
 
+    char keys_info[MAX_N_KEYS_IN_WALLET_POLICY][MAX_POLICY_KEY_INFO_LEN + 1];
+    key_type_e keys_type[MAX_N_KEYS_IN_WALLET_POLICY];
+    memset(keys_type, 0, sizeof(keys_type));
+
     for (size_t cosigner_index = 0; cosigner_index < wallet_header.n_keys; cosigner_index++) {
         /**
          * Receives and parses the next pubkey info.
          * Asks the user to validate the pubkey info.
          */
 
-        uint8_t next_pubkey_info[MAX_POLICY_KEY_INFO_LEN + 1];
-        int pubkey_info_len = call_get_merkle_leaf_element(dc,
-                                                           wallet_header.keys_info_merkle_root,
-                                                           wallet_header.n_keys,
-                                                           cosigner_index,
-                                                           next_pubkey_info,
-                                                           MAX_POLICY_KEY_INFO_LEN);
+        int key_info_len = call_get_merkle_leaf_element(dc,
+                                                        wallet_header.keys_info_merkle_root,
+                                                        wallet_header.n_keys,
+                                                        cosigner_index,
+                                                        (uint8_t *) keys_info[cosigner_index],
+                                                        MAX_POLICY_KEY_INFO_LEN);
 
-        if (pubkey_info_len < 0) {
+        if (key_info_len < 0) {
             SEND_SW(dc, SW_INCORRECT_DATA);
-            ui_post_processing_confirm_wallet_registration(dc, false);
             return;
         }
 
```

Comment:
I'd add a control on `key_info_len > MAX_POLICY_KEY_INFO_LEN + 1` even though I assume that `call_get_merkle_leaf_element` has already this control.